pkg/cacheutil/memcached_client.go

```diff
@@ -0,0 +1,382 @@
+package cacheutil
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/bradfitz/gomemcache/memcache"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/thanos-io/thanos/pkg/discovery/dns"
+)
+
+const (
+	defaultTimeout                     = 100 * time.Millisecond
+	defaultMaxIdleConnections          = 100
+	defaultMaxAsyncConcurrency         = 20
+	defaultMaxAsyncBufferSize          = 10000
+	defaultDNSProviderUpdateInterval   = 10 * time.Second
+	defaultMaxGetMultiBatchConcurrency = 20
+	defaultMaxGetMultiBatchSize        = 1024
+)
+
+var (
+	errMemcachedAsyncBufferFull = errors.New("the async buffer is full")
+	errMemcachedConfigNoAddrs   = errors.New("no memcached addrs provided")
+)
+
+// MemcachedClient is a high level client to interact with memcached.
+type MemcachedClient interface {
```

Comment:
> Do we need another interface?

The interface is currently used to mock it in `pkg/store/cache/memcached_test.go`. It may also be need to add multi-tenancy support in Cortex: an option may be having 1 single underlying memcached client and wrapping it through a proxy for each tenant having the proxy prefixing the keys with the tenant ID (but no decision has been taken and I'm open to suggestions).

> I think we way we do it is to accept interfaces but return structs as advised by Go (:

Right, done.

Reply:
It sounds like a YAGNI here. Does it mean creating distributing each client implementation for each block? (: Sounds like something we can add unless we know clear decision here, but let's discuss

Reply:
I'm not getting if the comment is about (a) removing the `MemcachedClient` interface or (b) merging `memcachedClient` with `MemcachedIndexCache`?